packages/web-react/src/components/TextFieldBase/TextFieldBase.tsx

```diff
@@ -1,29 +1,60 @@
-import classNames from 'classnames';
 import React from 'react';
-import { useStyleProps } from '../../hooks/styleProps';
+import classNames from 'classnames';
+import { useStyleProps } from '../../hooks';
 import { SpiritTextFieldBaseProps } from '../../types';
+import { Icon } from '../Icon';
+import { useTextFieldBase } from './useTextFieldBase';
 import { useTextFieldBaseStyleProps } from './useTextFieldBaseStyleProps';
 
 export const TextFieldBase = (props: SpiritTextFieldBaseProps): JSX.Element => {
   const { classProps, props: modifiedProps } = useTextFieldBaseStyleProps(props);
-  const { id, isDisabled, isMultiline, isRequired, label, inputWidth, message, ...restProps } = modifiedProps;
+  const { hasPasswordToggle, id, isDisabled, isMultiline, isRequired, label, inputWidth, message, type, ...restProps } =
+    modifiedProps;
   const { styleProps, props: otherProps } = useStyleProps(restProps);
+  const { inputType, passwordShown, passwordToggleHandler } = useTextFieldBase({
+    hasPasswordToggle,
+    isMultiline,
+    type,
+  });
 
   const ElementType: React.ElementType = isMultiline ? 'textarea' : 'input';
 
+  const input = (
+    <ElementType
+      {...otherProps}
+      className={classProps.input}
+      disabled={isDisabled}
+      id={id}
+      required={isRequired}
+      size={inputWidth}
+      type={inputType}
+    />
+  );
+
   return (
     <div {...styleProps} className={classNames(classProps.root, styleProps.className)}>
       <label htmlFor={id} className={classProps.label}>
         {label}
       </label>
-      <ElementType
-        {...otherProps}
-        id={id}
-        className={classProps.input}
-        size={inputWidth}
-        disabled={isDisabled}
-        required={isRequired}
-      />
+      {hasPasswordToggle ? (
+        <div className={classProps.passwordToggle}>
```

Comment:
Please, refactor this to `<TextFieldPasswordToggle ...>{input}</TextFieldPasswordToggle>` and create a separate component file for this. This should be a composition.

As @tomassychra has pointed out, this may not be the last adornment of the TextField component, see https://mui.com/material-ui/react-text-field/#input-adornments.
So having this as a separate component will be wise.

Reply:
I agree but I'm afraid it's beyond the scope of this issue. I will welcome the adornment/add-on concept in all technologies, not only in React 🙂.

Reply:
Ok, thanks for your reply. I have refactored this a little bit. So we should be prepared for the adornments in a near future.